src/transport/SecurePairingSession.cpp

```diff
@@ -56,6 +58,87 @@ SecurePairingSession::~SecurePairingSession()
     memset(&mKe[0], 0, sizeof(mKe));
 }
 
+typedef struct
+{
+    uint16_t mKeLen;
+    uint8_t mKe[kMAX_Hash_Length];
+    uint8_t mPairingComplete;
+    uint64_t mLocalNodeId;
+    uint64_t mPeerNodeId;
+    uint16_t mLocalKeyId;
+    uint16_t mPeerKeyId;
+    uint64_t padding;
+} SecurePairingSessionSerializable;
+
+CHIP_ERROR SecurePairingSession::Serialize(SecurePairingSessionSerialized & output)
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+
+    const NodeId localNodeId = (mLocalNodeId.HasValue()) ? mLocalNodeId.Value() : kUndefinedNodeId;
+    const NodeId peerNodeId  = (mPeerNodeId.HasValue()) ? mPeerNodeId.Value() : kUndefinedNodeId;
+    VerifyOrExit(CanCastTo<uint16_t>(mKeLen), error = CHIP_ERROR_INTERNAL);
+    VerifyOrExit(CanCastTo<uint64_t>(localNodeId), error = CHIP_ERROR_INTERNAL);
+    VerifyOrExit(CanCastTo<uint64_t>(peerNodeId), error = CHIP_ERROR_INTERNAL);
+    VerifyOrExit(CanCastTo<uint16_t>(sizeof(SecurePairingSessionSerializable)), error = CHIP_ERROR_INTERNAL);
+
+    {
+        uint8_t paired                                = (mPairingComplete) ? 1 : 0;
+        uint16_t keLen                                = static_cast<uint16_t>(mKeLen);
+        uint16_t serializedLen                        = 0;
+        SecurePairingSessionSerializable serializable = {
+            keLen, { 0 }, paired, localNodeId, peerNodeId, mLocalKeyId, mPeerKeyId, 0
+        };
+        memcpy(serializable.mKe, mKe, kMAX_Hash_Length);
+
+        VerifyOrExit(BASE64_ENCODED_LEN(sizeof(serializable)) <= sizeof(output.inner), error = CHIP_ERROR_INVALID_ARGUMENT);
```

Comment:
If it's `==`, will we still have space for our trailing `\0`?

Reply:
Line 98 (https://github.com/project-chip/connectedhomeip/pull/3284/files#diff-f9d17dd763aec77ae22f839d02b9e83c87a8e9c5036124d99f3c438b0df211c6R98) is ensuring that there's space for `\0`. This check is making sure that output buffer has enough space to serialize the b64 encoded data.

Reply:
Ah, yes, makes sense.